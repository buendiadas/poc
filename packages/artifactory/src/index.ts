export { BasicToken } from './codegen/BasicToken';
export { AmIRichAlready } from './codegen/AmIRichAlready';
