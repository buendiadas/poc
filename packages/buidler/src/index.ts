export * from './provider';
export * from './utils';
export * from './imports';
export * from './vm';
