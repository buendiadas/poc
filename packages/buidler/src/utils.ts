import fs from 'fs-extra';
import path from 'path';
import { BuidlerProvider } from './provider';
import {
  validateConfig,
  resolveConfig,
  defaultConfig,
  solcInputFilename,
  solcOutputFilename,
} from './imports';

export async function addCompilationResult(
  source: string,
  provider: BuidlerProvider
) {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const loaded = require(source);
  validateConfig(loaded);

  const config = resolveConfig(source, defaultConfig, loaded, []);
  const inputPath = path.join(config.paths.cache, solcInputFilename);
  const outputPath = path.join(config.paths.cache, solcOutputFilename);
  const compilerVersion = config.solc.version;

  const compilerInput = await fs.readJSON(inputPath, {
    encoding: 'utf8',
  });

  const compilerOutput = await fs.readJSON(outputPath, {
    encoding: 'utf8',
  });

  await provider.send('buidler_addCompilationResult', [
    compilerVersion,
    compilerInput,
    compilerOutput,
  ]);
}
