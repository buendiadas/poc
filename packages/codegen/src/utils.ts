export function prettierConfig(cwd: string = process.cwd()) {
  try {
    const prettier = require('prettier');
    return prettier.resolveConfig.sync(cwd);
  } catch (error) {
    return {};
  }
}

export function formatOutput(value: string) {
  try {
    const prettier = require('prettier');
    const defaults = prettierConfig();
    const options = {
      ...defaults,
      parser: 'typescript',
    };

    return prettier.format(value, options);
  } catch (error) {}

  return value;
}
