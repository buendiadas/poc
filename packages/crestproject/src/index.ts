import '@crestproject/jest/types';

export * from '@crestproject/ethers';
export * from '@crestproject/evm';
