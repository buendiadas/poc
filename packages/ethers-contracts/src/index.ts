export * from './construction';
export * from './contract';
export * from './function';
export * from './types';
