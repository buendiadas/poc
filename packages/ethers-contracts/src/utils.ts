export function propertyOf<TOr = any>(
  property: string,
  candidates: object[] = [],
): TOr {
  const obj = candidates.find((obj) => obj.hasOwnProperty(property));
  return (obj as any)?.[property] ?? undefined;
}
