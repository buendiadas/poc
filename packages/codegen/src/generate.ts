import { ConstructorFragment } from '@ethersproject/abi';
import { ethers } from 'ethers';
import { formatOutput } from './utils';

export function getInput(fragment: ConstructorFragment) {
  const inputs = fragment.inputs.map((input, index) => {
    const name = input.name || `$$${index}`;
    const type = getType(input, true);
    return `${name}: ${type}`;
  });

  return inputs.join(', ');
}

export function getOutput(fragment: ConstructorFragment) {
  if (!ethers.utils.FunctionFragment.isFunctionFragment(fragment)) {
    return 'void';
  }

  const outputs = fragment.outputs ?? [];
  if (!outputs.length) {
    return 'void';
  }

  if (outputs.length === 1) {
    return getType(outputs[0], false);
  }

  // If all output parameters are named and unique, we can specify the struct.
  const struct = outputs.every((item, index, array) => {
    if (!item.name) {
      return false;
    }

    const found = array.findIndex((inner) => inner.name === item.name);
    return found === index;
  });

  if (struct) {
    const kv = outputs.map((o) => `${o.name}: ${getType(o, false)}`);
    return `{ ${kv.join(', ')} }`;
  }

  // Otherwise, all we know is that it will be an array.
  return 'any[]';
}

export function getRawOutput(fragment: ConstructorFragment) {
  if (!ethers.utils.FunctionFragment.isFunctionFragment(fragment)) {
    return '[]';
  }

  const outputs = (fragment.outputs ?? []).map((output, index) => {
    const name = output.name || `$$${index}`;
    const type = getType(output, true);
    return `${name}: ${type}`;
  });

  return `[${outputs.join(', ')}]`;
}

export function getType(
  param: ethers.utils.ParamType,
  flexible?: boolean,
): string {
  if (param.type === 'array' || param.type.substr(-1) === ']') {
    const type = getType(param.arrayChildren, flexible);
    const matches = param.type.match(/\[([0-9]*)\]$/);

    if (matches?.[1]) {
      // This is a fixed length array.
      const range = Array.from(Array(parseInt(matches[1], 10)).keys());
      return `[${range.map(() => type).join(', ')}]`;
    }

    // Arbitrary length array.
    return `${type}[]`;
  }

  if (param.type === 'tuple') {
    const struct = param.components.map((param, index) => {
      return `${param.name || `$$${index}`}: ${getType(param, flexible)}`;
    });

    return `{ ${struct.join(', ')} }`;
  }

  if (param.type === 'address') {
    return flexible ? 'AddressLike' : 'string';
  }

  if (param.type === 'string') {
    return 'string';
  }

  if (param.type === 'bool') {
    return 'boolean';
  }

  if (param.type.substring(0, 5) === 'bytes') {
    return flexible ? 'ethers.utils.BytesLike' : 'string';
  }

  if (param.type.substring(0, 4) === 'uint') {
    return flexible ? 'ethers.BigNumberish' : 'ethers.BigNumber';
  }

  return 'any';
}

export function generateFunction(
  contract: string,
  fragment: ethers.utils.FunctionFragment,
) {
  const type = fragment.constant ? 'Call' : 'Send';
  const input = getInput(fragment);
  const output = getOutput(fragment);
  return `${type}<(${input}) => ${output}, ${contract}>`;
}

export function generateFunctions(
  contract: string,
  fragments: ethers.utils.FunctionFragment[],
) {
  if (!fragments.length) {
    return '';
  }

  const [short, full] = fragments.reduce(
    ([short, full], fragment, index, array) => {
      const type = generateFunction(contract, fragment);
      const found = array.findIndex(
        (current) => fragment.name === current.name,
      );

      // Only create a shortcut for the first function overload.
      if (index === found) {
        short.push(`${fragment.name}: ${type}`);
      }

      const signature = fragment.format();
      full.push(`'${signature}': ${type}`);

      return [short, full] as [string[], string[]];
    },
    [[], []] as [string[], string[]],
  );

  return `// Shortcuts (using function name of first overload)
  ${short.join('\n  ')}

  // Explicit accessors (using full function signature)
  ${full.join('\n  ')}`;
}

export function generateConstructorArgs(fragment: ConstructorFragment) {
  const input = getInput(fragment);
  return input ? `[${input}]` : '';
}

export function generateContractForSolidityArtifact(
  name: string,
  source: string,
  abi: ethers.utils.Interface,
  crestproject: string = '@crestproject/crestproject',
) {
  const functions = generateFunctions(name, Object.values(abi.functions));
  const constructor = generateConstructorArgs(abi.deploy);
  const generic = `${name}${constructor ? `, ${name}Args` : ''}`;

  // prettier-ignore
  return `/* eslint-disable */
// @ts-nocheck
import { ethers } from 'ethers';
import { contract, Call, Send, AddressLike, Contract } from '${crestproject}';
import ${name}Artifact from '${source}';

${constructor ? `export type ${name}Args = ${constructor};` : ''}

// prettier-ignore
export interface ${name} extends Contract<${name}> {
  ${functions || '// No external functions'}
}

export const ${name} = contract.fromArtifact<${generic}>(${name}Artifact);`;
}

export function generateContractForSignatures(
  name: string,
  abi: ethers.utils.Interface,
  crestproject: string = '@crestproject/crestproject',
) {
  const functions = generateFunctions(name, Object.values(abi.functions));
  const constructor = generateConstructorArgs(abi.deploy);
  const generic = `${name}${constructor ? `, ${name}Args` : ''}`;
  const formatted = abi.format();

  // prettier-ignore
  return `/* eslint-disable */
// @ts-nocheck
import { ethers } from 'ethers';
import { contract, Call, Send, AddressLike, Contract } from '${crestproject}';

${constructor ? `export type ${name}Args = ${constructor};` : ''}

// prettier-ignore
export interface ${name} extends Contract<${name}> {
  ${functions || '// No external functions'}
}

export const ${name} = contract.fromSignatures<${generic}>\`
  ${Array.isArray(formatted) ? formatted.join('\n  ') : formatted}
\`;`;
}
