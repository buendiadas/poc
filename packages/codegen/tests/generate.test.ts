import path from 'path';
import { utils } from 'ethers';
import { formatOutput } from '../src/utils';
import {
  generateContractForSolidityArtifact,
  generateContractForSignatures,
} from '../src/generate';

describe('code generator', () => {
  it('generates code for solidity artifacts', () => {
    const contract = require('./__fixtures__/artifact.json');
    const source = require.resolve('./__fixtures__/artifact.json');
    const destination = path.resolve(__dirname, 'ERC20.ts');

    let relative = path.relative(path.dirname(destination), source);
    if (!relative.startsWith('.')) {
      relative = `./${relative}`;
    }

    const abi = new utils.Interface(contract.abi);
    const output = generateContractForSolidityArtifact('ERC20', relative, abi);
    const formatted = formatOutput(output);

    expect(formatted).toMatchSnapshot();
  });

  it('generates code for ethers signatures', () => {
    const contract = require('./__fixtures__/artifact.json');
    const abi = new utils.Interface(contract.abi);
    const output = generateContractForSignatures('ERC20', abi);
    const formatted = formatOutput(output);

    expect(formatted).toMatchSnapshot();
  });
});
