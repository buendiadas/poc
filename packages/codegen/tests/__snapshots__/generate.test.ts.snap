// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code generator generates code for solidity artifacts 1`] = `
"/* eslint-disable */
// @ts-nocheck
import { BytesLike, BigNumber, BigNumberish } from 'ethers';
import {
  contract,
  Call,
  Send,
  AddressLikeAsync,
  Contract,
} from '@crestproject/crestproject';

export type ERC20Args = [name: string, symbol: string];

// prettier-ignore
export interface ERC20 extends Contract<ERC20> {
  // Shortcuts (using function name of first overload)
  allowance: Call<(owner: AddressLikeAsync, spender: AddressLikeAsync) => BigNumber, ERC20>
  approve: Send<(spender: AddressLikeAsync, amount: BigNumberish) => boolean, ERC20>
  balanceOf: Call<(account: AddressLikeAsync) => BigNumber, ERC20>
  decimals: Call<() => BigNumber, ERC20>
  decreaseAllowance: Send<(spender: AddressLikeAsync, subtractedValue: BigNumberish) => boolean, ERC20>
  increaseAllowance: Send<(spender: AddressLikeAsync, addedValue: BigNumberish) => boolean, ERC20>
  name: Call<() => string, ERC20>
  symbol: Call<() => string, ERC20>
  totalSupply: Call<() => BigNumber, ERC20>
  transfer: Send<(recipient: AddressLikeAsync, amount: BigNumberish) => boolean, ERC20>
  transferFrom: Send<(sender: AddressLikeAsync, recipient: AddressLikeAsync, amount: BigNumberish) => boolean, ERC20>

  // Explicit accessors (using full function signature)
  'allowance(address,address)': Call<(owner: AddressLikeAsync, spender: AddressLikeAsync) => BigNumber, ERC20>
  'approve(address,uint256)': Send<(spender: AddressLikeAsync, amount: BigNumberish) => boolean, ERC20>
  'balanceOf(address)': Call<(account: AddressLikeAsync) => BigNumber, ERC20>
  'decimals()': Call<() => BigNumber, ERC20>
  'decreaseAllowance(address,uint256)': Send<(spender: AddressLikeAsync, subtractedValue: BigNumberish) => boolean, ERC20>
  'increaseAllowance(address,uint256)': Send<(spender: AddressLikeAsync, addedValue: BigNumberish) => boolean, ERC20>
  'name()': Call<() => string, ERC20>
  'symbol()': Call<() => string, ERC20>
  'totalSupply()': Call<() => BigNumber, ERC20>
  'transfer(address,uint256)': Send<(recipient: AddressLikeAsync, amount: BigNumberish) => boolean, ERC20>
  'transferFrom(address,address,uint256)': Send<(sender: AddressLikeAsync, recipient: AddressLikeAsync, amount: BigNumberish) => boolean, ERC20>
}

let ERC20Bytecode: string | undefined = undefined;
if (typeof window === 'undefined') {
  ERC20Bytecode = './__fixtures__/artifact.json';
}

// prettier-ignore
export const ERC20 = contract<ERC20, ERC20Args>(ERC20Bytecode)\`
  constructor(string name, string symbol)
  event Approval(address indexed owner, address indexed spender, uint256 value)
  event Transfer(address indexed from, address indexed to, uint256 value)
  function allowance(address owner, address spender) view returns (uint256)
  function approve(address spender, uint256 amount) returns (bool)
  function balanceOf(address account) view returns (uint256)
  function decimals() view returns (uint8)
  function decreaseAllowance(address spender, uint256 subtractedValue) returns (bool)
  function increaseAllowance(address spender, uint256 addedValue) returns (bool)
  function name() view returns (string)
  function symbol() view returns (string)
  function totalSupply() view returns (uint256)
  function transfer(address recipient, uint256 amount) returns (bool)
  function transferFrom(address sender, address recipient, uint256 amount) returns (bool)
\`;
"
`;
